proposals/top-level-members.md

```diff
@@ -0,0 +1,142 @@
+  whose accessibility is either `internal` (by default) or `public` (if any member is also `public`).
+  For top-level members, this means:
+    - The `static` modifier is disallowed (the members are implicitly static).
+    - The default accessibility is `internal`.
+      `public` and `private` is also allowed.
+      `protected` and `file` is disallowed.
+    - Overloading is supported.
+    - `extern` and `partial` are supported.
+    - XML doc comments work.
+- Metadata:
+  - A type synthesized per namespace and file. That means `private` members are only visible in the file.
+  - Cannot be addressed from C#, but has speakable name `TopLevel` so it is callable from other languages.
+    This means that custom types named `TopLevel` become disallowed in a namespace where top-level members are used.
+  - It needs to have an attribute `[TopLevel]` otherwise it is considered a plain old type. This prevents a breaking change.
+- Usage (if there is an appropriately-shaped `NS.TopLevel` type):
+  - `using NS;` implies `using static NS.TopLevel;`.
+  - Lookup for `NS.Member` can find `NS.TopLevel.Member`.
+  - Nothing really changes for extensions.
+- Entry points:
+  - Top-level `Main` methods can be entry points.
+  - Top-level statements are generated into `Program.Main` (speakable function).
+    This is a breaking change (previously the main method was unspeakable).
+  - Simplify the logic: TLS entry-points are normal candidates.
+    This is a breaking change (previously they were not considered to be candidates and for example `-main` could not be used to point to them).
+
+## Drawbacks
+
+TODO: Why should we *not* do this?
+
+- Polluting namespaces with loosely organized helpers.
+- Requires tooling updates to properly surface and organize top-level methods in IntelliSense, refactorings, etc.
+- Entry point resolution breaking changes.
+
+## Alternatives
+
+TODO: What other designs have been considered? What is the impact of not doing this?
+
+- Support `args` keyword in top-level members (just like it can be accessed in top-level statements). But we have `System.Environment.GetCommandLineArgs()`.
+- Allow capturing variables from top-level statements inside non-`static` top-level members.
+  Could be used to refactor a single-file program into multi-file program just by extracting functions to separate files.
+  But it would mean that a method's implementation (top-level statements) can influence what other methods see (which variables are available in top-level members).
+- Allow declaring top-level members outside namespaces as well.
+  - Would introduce ambiguities with top-level statements.
+  - Could be brought to scope via `extern alias`.
+    - To avoid needing to specify those in project files (e.g., so file-based apps also work),
+      there could be a syntax for that like `extern alias Util = Util.dll`.
+
+## Open questions
+
+TODO: What parts of the design are still undecided?
```

Comment:
The rules for name conflicts should probably be spelled out.

For example, it seem as if the following declarations would be in conflict, due to how they are represented in the language

```cs
namespace NS;
int Foo;
class Foo { }
```